Automattic/wp-calypso
# Shopping cart: Fix cart key change and reload behavior

#### Changes proposed in this Pull Request

This fixes some bugs in the `@automattic/shopping-cart` package. None of the bugs here are causing any problems currently, but they will in the future as the package gains use. Particularly, this fixes the cart reload process where the cart will re-fetch its data from the server when the cart key changes or when manually requested to do so.

It also makes sure that the fake "wpcom credits" products generated by the shopping cart endpoint are stripped out before being passed along to consumers. They are not necessary because credits are available as a separate set of properties on the `ResponseCart` object and they require special exceptions in any code that iterates over products (eg: is the cart empty if it still contains a credits product?). Eventually there's an intent to remove these products from the backend as well, but for now we can at least make sure that frontend consumers don't have to deal with them.

#### Testing instructions

Verify that checkout is unchanged; add a product to your cart and visit checkout. Verify that there are no errors and that the cart product appears. Delete the product and verify that it goes away.

If you wish to test these new modifications themselves, you'll need to test with changes that use these features. https://github.com/Automattic/wp-calypso/pull/47605 requires them, so testing that PR will also validate these changes.

https://github.com/Automattic/wp-calypso/pull/47648 also adds integration tests for this fix.
